### DISCUSSION TITLE
Downloading SQLite DB from Web occasionally bugs out

### DISCUSSION THREAD

**Discussion by coogle** [simolus3/drift#2667] [Q&A]
I've got a problem that's hard to nail down...

The way I'm using drift is to build an SQLite database on a server, and then have the flutter app (web, specifically) download that database occasionally using `dio`, and write it to the IndexedDb VFS... from there drift can perform queries on it. The database is about 5.5MB in size.

Occasionally I am running into a problem where the new database doesn't download properly. When I look at the `filesize` attribute in the `files` for IndexedDb, sometimes for example it's only `65535`. Sometimes, it's showing the proper filesize but then SQLite says the tables don't exist... sometimes the VFS freaks out and starts constantly throwing exceptions about Array access being out of bounds (with an constantly decreasing negative length value).

Then, sometimes, it just works like it is intended to.

Needless to say I'm at a bit of a lost.

@simolus3 I'm sure I'm pushing the limits of the current capabilities here, so maybe you have some insight? Below is my code to write to the IndexedDb VFS.. is there anything here you are seeing that would cause this?

```
    Future<void> downloadNewDatabase({
        required Uri downloadUri,
        DatabaseDownloadProgressFn? progressCb,
        bool log = false
        }) async
    {
        Dio _dio = Dio();
        _dio.options.responseType = ResponseType.stream;

        if(kIsWeb)
        {
            _dio.httpClientAdapter = BrowserHttpClientAdapter();
        }

        if(log) 
        {
            _dio.interceptors.add(PrettyDioLogger(
                requestHeader: true,
                requestBody: true,
                responseBody: false,
                responseHeader: true,
            ));
        }

        final result = await _dio.get(
            downloadUri.toString(),
            onReceiveProgress: (actualBytes, totalBytes)
            {
                double percent = (actualBytes / totalBytes);

                if(progressCb != null)
                {
                    progressCb(percent);
                }
                
            }
        );

        if(kIsWeb)
        {
            _storeIndexedDbDatabase(result.data.stream);
        }
        else
        {
            _storeLocalDatabase(result.data.stream);
        }
    }
    
    Future<void> _storeIndexedDbDatabase(Stream<Uint8List> stream) async
    {
        IndexedDbFileSystem fs = await IndexedDbFileSystem.open(
            dbName: DatabaseDownloader.indexedDbName
        );

        if(fs.xAccess('/' + DatabaseDownloader.sqliteFilename, 0) != 0)
        {
            fs.xDelete('/' + DatabaseDownloader.sqliteFilename, 0);    
        }
        
        final idbFile = fs.xOpen(
            Sqlite3Filename('/' + DatabaseDownloader.sqliteFilename),
            SqlFlag.SQLITE_OPEN_CREATE | SqlFlag.SQLITE_OPEN_READWRITE
        );

        num offset = 0;
        await for(final bytes in stream)
        {
            idbFile.file.xWrite(bytes, offset.toInt());
            offset += bytes.length;
        }

        idbFile.file.xClose();

        await fs.flush();
    }
    ```

**Comment by simolus3** [comment]
From a quick look, the code looks correct to me (well, except that you'll probably want to await `_storeIndexedDbDatabase` and avoid opening a drift database while this operation is happening). Are there concurrent accesses to the same IndexedDB database? We don't currently support those.

Since an `IndexedDbFileSystem` holds the entire database in memory either way, you could try buffering the database into a `Uint8List` and store that in one `xWrite` invocation. If that fixes the problem, there'd still be a problem in the VFS implementation though. I can try to reproduce this by downloading a database with a similar size in a similar way. Are you downloading the database and opening it in drift directly afterwards?

**Reply by coogle** [reply]
There isn't any concurrent access here, I'm just basically experiencing this running local dev (`flutter run -d chrome`) with a single tab. 

Secondary question.. when you say you don't currently support concurrent accesses, does that also apply to read-only access? Not the immediate problem but just trying to think down the road when I have a user who opens up two tabs to the same Flutter app for some reason. Do you have any suggestions for how to work around that issue for a read-only DB like this?

Getting back to your question -- I create an instance of the database object itself early in the application (using `get_it`), but don't actually perform any queries on it immediately. From a flow perspective, the post-auth-success page is what executes the code I provided above to download the database... but the database itself isn't queried/accessed until one-screen deeper from that post-auth page.

When I get a moment I'll try your suggestion of consuming the full stream into a `Uint8List` and writing it all in one go, to see if it helps. That might take a bit only because at the moment I've got a functional DB and a pile of work that needs to be done requiring that DB before I can circle back to potentially breaking it again ;)

Thanks for looking into this!

**Reply by simolus3** [reply]
> does that also apply to read-only access?

Read-only access should be fine. You can also compile a shared web worker hosting the database ([doc link](https://drift.simonbinder.eu/web/#using-web-workers) if not using `WasmDatabase.open`) for concurrent (okay, serialized in the worker) writes.

> but the database itself isn't queried/accessed until one-screen deeper from that post-auth page.

That should be fine, creating the database object doesn't do much.

FWIW, you can essentially block the database from being opened by awaiting a completer in there (or just throwing if it's being opened too early):

```dart
LazyDatabase(() async {
  if (!authCompleted) { // or some better state management solution, you get the idea :)
    throw StateError('The database is not supposed to be used yet.');
  }

  await downloadNewDatabase(); // maybe here, or maybe await it before setting `authCompleted` to true
  return WasmDatabase(...);
})
```

It shouldn't make a difference, but if the database is accidentally opened early it's probably best to know. Opening the database and then later opening the same VFS again to write the downloaded database would lead to race conditions which could explain the symptoms you're seeing.

**Reply by coogle** [reply]
Agreed. I'll have to think about the best way to inject some sort of failsafe state checking into the connect, if only so that 6 mos from now I catch the regression. Especially if that might account for some of the symptoms I've had. 

Also -- good to know for read-only I wouldn't have to worry about the concurrency issue (which is what I hoped / expected).

**Comment by coogle** [comment]
An update on this issue, I still am ~~stuck~~ less stuck! What I've learned since I last posted..

At this point the only problem seems to be the `RangeError` I describe below I occasionally still get.

I tried doing what you suggested and buffering the entire download into a variable and doing a single call to `xWrite()` (In my case I had to buffer it into `List<int>` and then call `Uint8List.fromList(myList)` because I didn't know the size of the database ahead of time). Either way, this had no effect as I still ran into `RangeError` exceptions that were constantly being thrown. Here is one of them for your reference:

```
RangeError: Invalid typed array length: -5267456
dart-sdk/lib/_internal/js_dev_runtime/private/native_typed_data.dart 1077:7   _create3
dart-sdk/lib/_internal/js_dev_runtime/private/native_typed_data.dart 1051:11  view
dart-sdk/lib/_internal/js_dev_runtime/private/native_typed_data.dart 34:28    asUint8List]
packages/sqlite3/src/wasm/vfs/indexed_db.dart 168:38                          <fn>
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 45:50            <fn>
dart-sdk/lib/async/zone.dart 1407:47                                          _rootRunUnary
dart-sdk/lib/async/zone.dart 1308:19                                          runUnary
dart-sdk/lib/async/future_impl.dart 156:18                                    handleValue
dart-sdk/lib/async/future_impl.dart 840:44                                    handleValueCallback
dart-sdk/lib/async/future_impl.dart 869:13                                    _propagateToListeners
dart-sdk/lib/async/future_impl.dart 641:5                                     [_completeWithValue]
dart-sdk/lib/async/future_impl.dart 715:7                                     <fn>
dart-sdk/lib/async/zone.dart 1399:13                                          _rootRun
dart-sdk/lib/async/zone.dart 1301:19                                          run
dart-sdk/lib/async/zone.dart 1233:23                                          callback
dart-sdk/lib/async/schedule_microtask.dart 40:11                              _microtaskLoop
dart-sdk/lib/async/schedule_microtask.dart 49:5                               _startMicrotaskLoop
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 181:15           <fn>
```

Update: The below turned out to be a bug on my end -- specifically what was happening is what you first suggested which I was opening the database earlier than I realized. My DB update logic actually checks the timestamp in the local database to decide if an update is necessary, thus I was opening the database -- downloading a new database -- and not closing the original connection which was causing problems. 

The fix was, upon updating the database, rebuild the connection. Since I was using a singleton via `get_it` for this connection it was just a matter of resetting the lazy-init singleton I created after the new database was downloaded. 

~~The above is when it just doesn't seem to work at all.. When it does look like it's going to work, I'm still having problems:~~

~~I also implemented a SHA256 hash on both the server and client side. Now the server when it builds the SQLite DB also calculates a SHA256 of it, which it then passes along to the Flutter app with the download URL. I then in the flutter app calculate the SHA256 as I am writing it to the IndexedDb VFS and at the conclusion make sure it matches.~~

~~Despite the fact that the SHA256 matches after I've written it to the VFS, when I actually go to actually use the database I can't find the tables and get errors like this:~~

~~`SqliteException(1): while preparing statement, no such table: sites`~~

~~As was with things before -- sometimes this DOES work exactly as expected, but it often requires deleting the IndexedDB and reloading multiple times before it finally manages to get a decent database.~~

@simolus3 any further thoughts or information I can help provide?

**Comment by coogle** [comment]
More updates!

I don't know *why* it's happening but I've gotten a step closer to identifying why, when the download seems to work, I end up with a database that has no tables in it. Here's what I'm observing:

1) I fetch the database from the remote server and it works (no `RangeError` exceptions)
2) I look in the Devtools at the `IndexedDb` at the `fileName` entry and I see my `/frontend.sqlite` database with the appropriate file size (~ 5MB)
3) I then click forward through the application to where I perform my first queries against the database, and I get an error about my table not being found
4) When I refresh the IndexedDb database, the filesize has dropped from ~5MB to 8192 bytes! 

So it seems like there is something happening here where the database I downloaded is being overwritten with an empty database.  

I've noted the following: If I download the database successfully (Step 1), skip Step 3 (try to use it), and instead hot-reload the Flutter application... on the 2nd run through it doesn't try to download the database again (as expected), and it doesn't klobber the database when I go to use it at Step 4. This implies to me there is something *not* happening once I write the database to the VFS causing the klobbering issue -- something that happens automatically when I hot reload (e.g. maybe a resource isn't being closed and reopened that needs it?)

Does that help at all?

**Reply by simolus3** [reply]
Are you calling `IndexedDbFileSystem.open()` in a `LazyDatabase` when opening the database? That method eagerly creates an in-memory view of the IndexedDB state when called, so if it's not wrapped in a `LazyDatabase` to delay it, what might be happening is that:

1. To construct the database (without using it), you're calling `IndexedDbFileSystem.open`.
2. You're then downloading a new database file.
3. The database uses the VFS created in step 1, which has an outdated state in memory.

We unfortunately need an in-memory copy in the IndexedDB VFS since sqlite3 expects to access file systems synchronously. But if the copy is created too early,
